### PR TITLE
Test for vector-valued ZhangLi u

### PR DESCRIPTION
**Pull request by lang-m**
This is related to ubermag/micromagneticmodel#75 and ubermag/oommfc#133

Merging this test will also require some changes in mumax3c: either we can support the same functionality or disable the test. I will try that when I am on a suitable machine.

### CODE REVIEW

**Comment by lang-m** [comment]
@samjrholt Can you please review the new test `test_vector_scalar`.

**Comment by lang-m** [comment]
I re-wrote all tests for Zhang Li, the diff is not really meaningful. The new tests all check some sort of DW motion.

Testing the new methods with mumax3c remains to be done.

**Comment by lang-m** [comment]
Support in mumax3c is implemented in ubermag/mumax3c#88

File: micromagnetictests/calculatortests/zhangli.py [hunk 1]
```diff
@@ -1,252 +1,306 @@
...
-        system.dynamics = mm.ZhangLi(u=u, beta=beta)
-        system.m = df.Field(mesh, nvdim=3, value=(0, 0.1, 1), norm=Ms)
+        system = mm.System(name="strip_x")
+        system.energy = mm.Exchange(A=A) + mm.UniaxialAnisotropy(K=K, u=anisotropy_axis)
+        mesh = df.Mesh(p1=p1, p2=p2, cell=cell, subregions=subregions)
+        system.m = df.Field(mesh, nvdim=3, value=init_m, norm=Ms)
 
-        td = self.calculator.TimeDriver()
-        td.drive(system, t=0.2e-9, n=50)
+        md = self.calculator.MinDriver()
```
Comment (samjrholt) [inline micromagnetictests/calculatortests/zhangli.py:47]: Is this okay for Mumax3?

File: micromagnetictests/calculatortests/zhangli.py [hunk 2]
```diff
@@ -1,252 +1,306 @@
...
+        # ensure that the initial state is as expected,
+        # a domain wall at y≈70 nm
+        assert 0.2 < system.m.orientation.z.mean() < 0.4
+        assert system.m.orientation.z.sel(y=(0, 60e-9)).mean() < -0.95
+        assert system.m.orientation.z.sel(y=(80e-9, 200e-9)).mean() > 0.95
+
+        system.dynamics = (
+            mm.Precession(gamma0=mm.consts.gamma0)
+            + mm.Damping(alpha=0.3)
+            + mm.ZhangLi(u=(0, 200, 0), beta=0.5)
```
Comment (samjrholt) [inline micromagnetictests/calculatortests/zhangli.py:297]:
```suggestion
            + mm.ZhangLi(u=(0, self.u, 0), beta=self.beta)
```

**Review by samjrholt — CHANGES_REQUESTED** [review]
Looks good in general. Should we also have a test with a vector field? I think that is the only case which isn't covered at the moment.

File: micromagnetictests/calculatortests/zhangli.py (hunk 1 above)
Comment (lang-m) [reply micromagnetictests/calculatortests/zhangli.py:47]: @samjrholt Would you prefer more local tests before merging this?

**Comment by samjrholt** [comment]
> Looks good in general. Should we also have a test with a vector field? I think that is the only case which isn't covered at the moment.

@lang-m did you implement this or decide it wasn't needed?

**Comment by lang-m** [comment]
There was a test already: `test_field_vector_u`. Current is just in x direction but that should not matter because we mainly need to tests our conversion code.

**Pull request merged**